src/mbi/relaxed_projection.py

```diff
@@ -0,0 +1,368 @@
+
+        result = []
+        for attr in self.domain.attrs:
+            p = sampled_data[:, self.feature_indices_map[attr]]
+            key, sk = jax.random.split(key)
+            ids = jax.random.categorical(sk, jnp.log(jnp.clip(p, a_min=1e-30, a_max=1.0)), axis=-1)
+            result.append(ids)
+
+        sampled_df = pd.DataFrame(np.stack([np.array(r) for r in result], axis=1), columns = list(self.domain.attrs))
+        sampled_data = Dataset(sampled_df, self.domain)
+        return sampled_data
+
+
+def relaxed_projection_estimation(
+    domain: Domain,
+    measurements: List[tuple],
+    iters: int = 1000,
+    lr: float = 0.01,
+    D_start: jax.Array = None,
+    seed: int = 0,
+    D_size: int = 1000,
+) -> ProjectableData:
+    """
+    API function for relaxed projection mechanism, return a 
+    ProjectableData class, which can further be used to conduct 
+    projection and synthesize data.
+
+    Args:
+        domain: a Domain class.
+        measurements: a list of tuples, each of which should be in 
+                      form ``(clique, measurement of clique)``.
+        iters: number of iterations in training.
+        lr: learning rate in training.
+        D_start: initial data on which we start the optimization. If 
+                 None, the optimization will start from a randomly 
+                 initialized data.
+        seed: random seed.
+        D_size: number of records in optimizable data.
+    """
+    key = jax.random.PRNGKey(seed)
+    if D_start is None:
+        D_start = _initialize_synthetic_dataset(key, num_generated_points=D_size, data_dimension=np.sum(domain.shape))
+
+    stat_dim = _obtain_dim(measurements = measurements)
+    statistics = [MarginalStatistics(domain, dim) for dim in stat_dim]
+    selected_query_index, measured_ans = _marginal_stat(measurements, stat_dim, statistics)
+
+    exact_fn, diff_fn = [None]*len(stat_dim), [None]*len(stat_dim)
+    for i in range(len(stat_dim)):
+        exact_fn[i] = statistics[i].get_exact_statistics_fn()
+        diff_fn[i]  = statistics[i].get_differentiable_statistics_fn()
+
+    @jax.jit
+    def progress_loss(D_prime, query_idx, target, _exact_fn=exact_fn):
+        loss = 0.0
+        for i in range(len(stat_dim)):
+            stats = _exact_fn[i](D_prime, query_idx[i])
+            loss += jnp.linalg.norm(target[i] - stats)**2
+        return loss
+    prog_fn = jax.jit(lambda D: progress_loss(D, selected_query_index, measured_ans))
+
+    @jax.jit
+    def loss_fn(D_prime, query_idx, target, _diff_fn=diff_fn):
+        loss = 0.0
+        for i in range(len(stat_dim)):
+            stats = _diff_fn[i](D_prime, query_idx[i])
+            loss += jnp.linalg.norm(target[i] - stats)**2
+        return loss
+
+    feats_cum = jnp.array([0] + list(domain.shape)).cumsum()
+    feats_idx = [list(range(feats_cum[i], feats_cum[i+1])) for i in range(len(feats_cum)-1)]
+    optimizer = optax.adam(learning_rate=lr)
```

Comment:
rather than hard-code adam and expose lr as the kwarg, it might be nice to make the optimizer be the kwarg and have it default to optax.adam(learning_rate=0.01), that makes it easier to configure the code to use different optimizers